### PR TITLE
Refactoring IServerFactory #395

### PR DESCRIPTION
**Pull request by JunTaoLuo**
#395

### CODE REVIEW

File: src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs [hunk 1]
```diff
@@ -245,10 +245,10 @@ private void EnsureServer()
...
             }
 
-            if (_serverFeatures == null)
+            if (_server == null)
             {
-                _serverFeatures = ServerFactory.Initialize(_config);
-                var addresses = _serverFeatures?.Get<IServerAddressesFeature>()?.Addresses;
+                _server = ServerFactory.CreateServer(_config);
+                var addresses = _server.Features?.Get<IServerAddressesFeature>()?.Addresses;
                 if (addresses != null && !addresses.IsReadOnly)
```
Comment (cesarbs) [inline src/Microsoft.AspNet.Hosting/Internal/HostingEngine.cs]: nit: you could do if (!addresses?.IsReadOnly)

File: src/Microsoft.AspNet.Hosting.Server.Abstractions/project.json [hunk 2]
```diff
@@ -6,6 +6,7 @@
     "url": "git://github.com/aspnet/hosting"
   },
   "dependencies": {
+    "Microsoft.AspNet.Http.Abstractions": "1.0.0-*",
     "Microsoft.AspNet.Http.Features": "1.0.0-*",
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Hosting.Server.Abstractions/project.json]: You can remove Features, it's transitive.

**Comment by Tratcher** [comment]
You're off to a good start. Next is moving IHttpContextFactory to Http.Abstractions and HttpContextFactory to Http. Then you need to update TestHost, Kestrel, and Entropy.

File: src/Microsoft.AspNet.Hosting.Server.Abstractions/IServer.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNet.Builder;
+using Microsoft.AspNet.Http.Features;
+
+namespace Microsoft.AspNet.Hosting.Server
+{
+    public interface IServer : IDisposable
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting.Server.Abstractions/IServer.cs]: Add some doc comments

File: src/Microsoft.AspNet.Hosting.Server.Abstractions/IServerFactory.cs [hunk 4]
```diff
@@ -1,16 +1,12 @@
...
-using System;
-using System.Threading.Tasks;
-using Microsoft.AspNet.Http.Features;
 using Microsoft.Extensions.Configuration;
 
 namespace Microsoft.AspNet.Hosting.Server
 {
     public interface IServerFactory
     {
-        IFeatureCollection Initialize(IConfiguration configuration);
```
Comment (davidfowl) [inline src/Microsoft.AspNet.Hosting.Server.Abstractions/IServerFactory.cs]: Same here, add some doc comments

File: test/Microsoft.AspNet.Hosting.Tests/HostingEngineTests.cs [hunk 5]
```diff
@@ -26,9 +27,8 @@
 
 namespace Microsoft.AspNet.Hosting
 {
-    public class HostingEngineTests : IServerFactory
+    public class HostingEngineTests
```
Comment (davidfowl) [inline test/Microsoft.AspNet.Hosting.Tests/HostingEngineTests.cs]: To reduce the test churn, make this type implement both IServer and IServerFactory

File: src/Microsoft.AspNet.TestHost/TestServer.cs [hunk 6]
```diff
@@ -14,18 +14,20 @@
 
 namespace Microsoft.AspNet.TestHost
 {
-    public class TestServer : IServerFactory, IDisposable
+    public class TestServer : IDisposable
```
Comment (davidfowl) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: Same comment here, make this implement both to reduce churn

File: src/Microsoft.AspNet.TestHost/TestServer.cs [hunk 7]
```diff
@@ -14,18 +14,20 @@
...
     {
         private const string DefaultEnvironmentName = "Development";
         private const string ServerName = nameof(TestServer);
-        private static readonly IFeatureCollection ServerInfo = new FeatureCollection();
-        private Func<IFeatureCollection, Task> _appDelegate;
+        private RequestDelegate _appDelegate;
         private IDisposable _appInstance;
         private bool _disposed = false;
+        private IHttpContextAccessor _httpContextAccessor;
```
Comment (davidfowl) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: This looks wrong, you shouldn't have to do any of this. It's using the WebHostBuilder under the covers which does the right thing.

File: src/Microsoft.AspNet.TestHost/ClientHandler.cs [hunk 8]
```diff
@@ -121,6 +124,12 @@ internal RequestState(HttpRequestMessage request, PathString pathBase)
                 }
 
                 HttpContext = new DefaultHttpContext();
+
+                if (accessor != null)
```
Comment (davidfowl) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: This also looks wrong.

File: src/Microsoft.AspNet.TestHost/ClientHandler.cs [hunk 9]
```diff
@@ -23,14 +24,15 @@ namespace Microsoft.AspNet.TestHost
     /// </summary>
     public class ClientHandler : HttpMessageHandler
```
Comment (davidfowl) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: This type shouldn't have changed

File: src/Microsoft.AspNet.Hosting/Builder/HttpContextFactory.cs [hunk 10]
```diff
@@ -1,26 +0,0 @@
-// Copyright (c) .NET Foundation. All rights reserved.
-// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
-
-using Microsoft.AspNet.Http;
-using Microsoft.AspNet.Http.Features;
-using Microsoft.AspNet.Http.Internal;
-
-namespace Microsoft.AspNet.Hosting.Builder
-{
-    public class HttpContextFactory : IHttpContextFactory
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Hosting/Builder/HttpContextFactory.cs]: Make a separate PR to fix the build ASAP.

File: src/Microsoft.AspNet.TestHost/TestServer.cs (hunk 7 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: No, IHttpContextAccessor usage was moved into IHttpContextFactory which isn't useful to TestServer since it prefers to create the DefaultHttpContext directly.

File: src/Microsoft.AspNet.TestHost/TestServer.cs [hunk 11]
```diff
@@ -14,18 +14,20 @@
...
 
         public TestServer(WebHostBuilder builder)
         {
-            _appInstance = builder.UseServer(this).Build().Start();
+            var hostingEngine = builder.UseServer(new ServerFactory(requestDelegate => { _appDelegate = requestDelegate; })).Build();
```
Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: Would it simplify matters if you added a UserServer overload that took an IServer directly? Then you could bypass the ServerFactory layer.

File: src/Microsoft.AspNet.Hosting/Builder/HttpContextFactory.cs (hunk 10 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNet.Hosting/Builder/HttpContextFactory.cs]: Thanks for doing this yesterday. I was hoping to do everything in one go but that was overzealous.

File: src/Microsoft.AspNet.TestHost/TestServer.cs (hunk 7 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: We should just use the default factory

File: src/Microsoft.AspNet.TestHost/ClientHandler.cs (hunk 8 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: I'm assuming this is because we need to decide whether to use the default factory or keep this hack for the custom factory?

File: src/Microsoft.AspNet.TestHost/ClientHandler.cs (hunk 9 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: What's wrong here?

File: src/Microsoft.AspNet.TestHost/TestServer.cs (hunk 11 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: I can. What's the order of precedence here? Look for the server, if not found try to create one with the resolved serverfactory?

**Comment by davidfowl** [comment]
Based on what Chris said. We should update we host builder and test server to support taking and IServer or ISeverFactory. There's probably an effect on the restartability

File: src/Microsoft.AspNet.TestHost/ClientHandler.cs (hunk 9 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: We should be using the default http context factory that sets the accessor for us automatically.

File: test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs [hunk 12]
```diff
@@ -40,60 +41,65 @@ public void Build_honors_UseStartup_with_string()
         [Fact]
         public async Task StartupMissing_Fallback()
         {
+            var application = new RequestDelegate(context => { return Task.FromResult(0); });
```
Comment (davidfowl) [inline test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs]: Why the ceremony here? Shouldn't this just look like it did before?

Comment (JunTaoLuo) [inline test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs]: This will go away once `UseServer` takes an `IServer`. This hack was added so the `RequestDelegate` passed into the `IServer` could be accessed for use in `AssertResponseContains`.

Comment (JunTaoLuo) [inline test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs]: Alternatively, we can just make this test extend `IServer` and `IServerFactory` like `HostingEngineTests`.

File: src/Microsoft.AspNet.TestHost/ClientHandler.cs (hunk 9 above)
Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: The default http context factory is less convienet to use here, but it would make it so you don't have to think about the accessor.

File: src/Microsoft.AspNet.TestHost/TestServer.cs [hunk 13]
```diff
@@ -14,22 +14,26 @@
 
 namespace Microsoft.AspNet.TestHost
 {
-    public class TestServer : IServerFactory, IDisposable
+    public class TestServer : IServerFactory, IServer, IDisposable
```
Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: IDisposable is now redundant.

File: src/Microsoft.AspNet.TestHost/TestServer.cs [hunk 14]
```diff
@@ -14,22 +14,26 @@
...
         {
-            _appInstance = builder.UseServer(this).Build().Start();
+            var hostingEngine = builder.UseServer(this).Build();
+            _httpContextAccessor = hostingEngine.ApplicationServices.GetService<IHttpContextAccessor>();
+            _appInstance = hostingEngine.Start();
         }
 
         public Uri BaseAddress { get; set; } = new Uri("http://localhost/");
 
+        public IFeatureCollection Features { get; }
```
Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: Some of these interface members don't actually make sense here, use the explicit implementation to hide them:
`IServer.Features { get; }`

Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: Is this ever set?

File: src/Microsoft.AspNet.TestHost/TestServer.cs [hunk 15]
```diff
@@ -145,5 +137,15 @@ public void Dispose()
             _disposed = true;
             _appInstance.Dispose();
         }
+
+        public IServer CreateServer(IConfiguration configuration)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: Explicit

File: src/Microsoft.AspNet.TestHost/TestServer.cs [hunk 16]
```diff
@@ -145,5 +137,15 @@ public void Dispose()
...
+        {
+            return this;
+        }
+
+        public void Start(RequestDelegate requestDelegate)
```
Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/TestServer.cs]: explicit

File: src/Microsoft.AspNet.TestHost/WebSocketClient.cs [hunk 17]
```diff
@@ -94,8 +94,7 @@ public RequestState(Uri uri, PathString pathBase, CancellationToken cancellation
                 _clientWebSocketTcs = new TaskCompletionSource<WebSocket>();
 
                 // HttpContext
-                FeatureCollection = new FeatureCollection();
-                HttpContext = new DefaultHttpContext(FeatureCollection);
+                HttpContext = new DefaultHttpContext(new FeatureCollection());
```
Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/WebSocketClient.cs]: I guess this should also use the context factory.

File: src/Microsoft.AspNet.TestHost/ClientHandler.cs (hunk 9 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: Yah, so use it

File: src/Microsoft.AspNet.TestHost/ClientHandler.cs (hunk 8 above)
Comment (davidfowl) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: Use the default factory everywhere

File: test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs [hunk 18]
```diff
@@ -125,36 +125,31 @@ private WebHostBuilder CreateWebHostBuilder()
...
             httpContext.Response.Body.Seek(0, SeekOrigin.Begin);
             var bodyText = new StreamReader(httpContext.Response.Body).ReadToEnd();
             Assert.Contains(expectedText, bodyText);
         }
 
-        private class TestServerFactory : IServerFactory
+        private class TestServer : IServer
         {
-            public Func<IFeatureCollection, Task> Application { get; set; }
+            private Action<RequestDelegate> _setAppDelegate;
```
Comment (JunTaoLuo) [inline test/Microsoft.AspNet.Hosting.Tests/WebHostBuilderTests.cs]: removing this

**Comment by JunTaoLuo** [comment]
updated

File: src/Microsoft.AspNet.Hosting.Server.Abstractions/IServerFactory.cs [hunk 19]
```diff
@@ -1,16 +1,20 @@
...
-        IDisposable Start(IFeatureCollection serverFeatures, Func<IFeatureCollection, Task> application);
+        /// <summary>
+        /// Creates <see cref="IServer"/> based on the given configuration.
+        /// </summary>
+        /// <param name="configuration">An instance of <see cref="IConfiguration"/>.</param>
+        /// <returns>The created server.</returns>
+        IServer CreateServer(IConfiguration configuration);
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Hosting.Server.Abstractions/IServerFactory.cs]: Create or CreateServer?

File: src/Microsoft.AspNet.Hosting/WebHostBuilder.cs [hunk 20]
```diff
@@ -172,7 +185,7 @@ public WebHostBuilder UseServer(string assemblyName)
             return this;
         }
 
-        public WebHostBuilder UseServer(IServerFactory factory)
+        public WebHostBuilder UseServerFactory(IServerFactory factory)
         {
             _serverFactory = factory;
```
Comment (Tratcher) [inline src/Microsoft.AspNet.Hosting/WebHostBuilder.cs]: Add a null check here for consistency?

File: src/Microsoft.AspNet.TestHost/ClientHandler.cs [hunk 21]
```diff
@@ -120,7 +130,8 @@ internal RequestState(HttpRequestMessage request, PathString pathBase)
                     request.Headers.Host = request.RequestUri.GetComponents(UriComponents.HostAndPort, UriFormat.UriEscaped);
                 }
 
-                HttpContext = new DefaultHttpContext();
+                HttpContext = factory.Create(new FeatureCollection());
```
Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: nit: contextFactory

Comment (Tratcher) [inline src/Microsoft.AspNet.TestHost/ClientHandler.cs]: _contextFactory

File: src/Microsoft.AspNet.Hosting.Server.Abstractions/IServerFactory.cs (hunk 19 above)
Comment (JunTaoLuo) [inline src/Microsoft.AspNet.Hosting.Server.Abstractions/IServerFactory.cs]: According to the issue filed, CreateServer.

**Comment by Tratcher** [comment]
:shipit: (once we get the servers updated)

**Comment by davidfowl** [comment]
This looks great :shipit:
